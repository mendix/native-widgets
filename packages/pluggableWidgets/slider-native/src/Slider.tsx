import { available, flattenStyles, toNumber, unavailable } from "@mendix/piw-native-utils-internal";
import { executeAction } from "@mendix/piw-utils-internal";
import MultiSlider, { MarkerProps } from "@ptomasroos/react-native-multi-slider";
import { createElement, ReactElement, useCallback, useRef, useState } from "react";
import { LayoutChangeEvent, Text, View } from "react-native";
import { Big } from "big.js";

import { SliderProps } from "../typings/SliderProps";
import { Marker } from "./Marker";
import { defaultSliderStyle, SliderStyle } from "./ui/Styles";

export type Props = SliderProps<SliderStyle>;

export function Slider(props: Props): ReactElement {
    const [width, setWidth] = useState<number>();

    const lastValue = useRef<number | undefined>(toNumber(props.valueAttribute));

    const value = toNumber(props.valueAttribute);
    const validationMessages = validate(props);
    const validProps = validationMessages.length === 0;
    const editable = props.editable !== "never" && !props.valueAttribute.readOnly && validProps;
    const styles = flattenStyles(defaultSliderStyle, props.style);

    const customMarker: Function = () => (markerProps: MarkerProps): JSX.Element => (
        <Marker {...markerProps} testID={`${props.name}$marker`} />
    );

    const onLayout = useCallback((event: LayoutChangeEvent): void => {
        setWidth(event.nativeEvent.layout.width);
    }, []);

    const onSlide = useCallback(
        (values: number[]): void => {
            if (values[0] === null) {
                return;
            }
            props.valueAttribute.setValue(new Big(values[0]));
        },
        [props.valueAttribute]
    );

    const onChange = useCallback(
        (values: number[]): void => {
            if (values[0] === null || lastValue.current === values[0]) {
                return;
            }

            lastValue.current = values[0];
            props.valueAttribute.setValue(new Big(values[0]));

            executeAction(props.onChange);
        },
        [lastValue.current, props.valueAttribute, props.onChange]
    );

    return (
        <View onLayout={onLayout} style={styles.container} testID={props.name}>
            <MultiSlider
                values={value != null ? [value] : undefined}
                min={validProps ? toNumber(props.minimumValue) : undefined}
                max={validProps ? toNumber(props.maximumValue) : undefined}
                step={validProps ? toNumber(props.stepSize) : undefined}
                enabledOne={editable}
                markerStyle={editable ? styles.marker : styles.markerDisabled}
                trackStyle={editable ? styles.track : styles.trackDisabled}
                selectedStyle={editable ? styles.highlight : styles.highlightDisabled}
                pressedMarkerStyle={styles.markerActive}
                onValuesChange={onSlide}
                onValuesChangeFinish={onChange}
                sliderLength={width}
                allowOverlap
                customMarker={customMarker()}
            />
            {!validProps && <Text style={styles.validationMessage}>{validationMessages.join("\n")}</Text>}
            {props.valueAttribute.validation && (
                <Text style={styles.validationMessage}>{props.valueAttribute.validation}</Text>
            )}
        </View>
    );
}

function validate(props: Props): string[] {
    const messages: string[] = [];
    const { minimumValue, maximumValue, stepSize, valueAttribute } = props;

    if (unavailable(minimumValue)) {
        messages.push("No minimum value provided.");
    }
    if (unavailable(maximumValue)) {
        messages.push("No maximum value provided.");
    }
    if (unavailable(stepSize)) {
        messages.push("No step size provided.");
    }
    if (unavailable(valueAttribute)) {
        messages.push("The value attribute is not readable.");
    }
    if (available(minimumValue) && available(maximumValue) && available(stepSize) && available(valueAttribute)) {
        if (stepSize.value!.lte(0)) {
            messages.push("The step size can not be zero or less than zero.");
        }
        if (minimumValue.value!.gt(maximumValue.value!)) {
            messages.push("The minimum value can not be greater than the maximum value.");
        } else {
            if (minimumValue.value!.eq(maximumValue.value!)) {
                messages.push("The minimum value can not be equal to the maximum value.");
            }
            if (valueAttribute.value!.lt(minimumValue.value!)) {
                messages.push("The current value can not be less than the minimum value.");
            }
            if (valueAttribute.value!.gt(maximumValue.value!)) {
                messages.push("The current value can not be greater than the maximum value.");
            }
        }
    }

    return messages;
}
