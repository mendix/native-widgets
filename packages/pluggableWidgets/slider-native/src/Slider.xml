<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.slider.Slider" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Slider</name>
    <description>Change a numeric value using a slider.</description>
    <studioProCategory>Input elements</studioProCategory>
    <studioCategory>Input elements</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Accessibilty">
                <property key="accessible" type="enumeration" defaultValue="yes">
                    <caption>Accessible</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="yes">Yes</enumerationValue>
                        <enumerationValue key="no">No</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="screenReaderCaption" type="textTemplate" required="false">
                    <caption>Screen reader caption</caption>
                    <description />
                </property>
                <property key="screenReaderHint" type="textTemplate" required="false">
                    <caption>Screen reader hint</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Data source">
                <property key="valueAttribute" type="attribute">
                    <caption>Value attribute</caption>
                    <description />
                    <attributeTypes>
                        <attributeType name="Integer" />
                        <attributeType name="Long" />
                        <attributeType name="Decimal" />
                    </attributeTypes>
                </property>
            </propertyGroup>
            <propertyGroup caption="Editability">
                <property key="editable" type="enumeration" defaultValue="default">
                    <caption>Editable</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="default">Default</enumerationValue>
                        <enumerationValue key="never">Never</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Range">
                <property key="minimumValue" type="expression" defaultValue="0">
                    <caption>Minimum value</caption>
                    <description>The minimum value that can be selected.</description>
                    <returnType type="Decimal" />
                </property>
                <property key="maximumValue" type="expression" defaultValue="100">
                    <caption>Maximum value</caption>
                    <description>The maximum value that can be selected.</description>
                    <returnType type="Decimal" />
                </property>
                <property key="stepSize" type="expression" defaultValue="1">
                    <caption>Step size</caption>
                    <description>The interval between two selectable points.</description>
                    <returnType type="Decimal" />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Events">
            <property key="onChange" type="action" required="false">
                <caption>On change</caption>
                <description>Called when the user releases the handle and the value has changed.</description>
            </property>
        </propertyGroup>
    </properties>
</widget>
