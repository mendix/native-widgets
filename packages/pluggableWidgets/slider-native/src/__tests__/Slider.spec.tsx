import { actionValue, dynamicValue, EditableValueBuilder } from "@mendix/piw-utils-internal";
import { Big } from "big.js";
import { createElement } from "react";
import { View } from "react-native";
import { fireEvent, render, RenderAPI } from "@testing-library/react-native";
import { ReactTestInstance } from "react-test-renderer";
import { ValueStatus, DynamicValue } from "mendix";

import { Props, Slider } from "../Slider";

describe("Slider", () => {
    const noValue: DynamicValue<Big> = { status: ValueStatus.Unavailable, value: undefined };
    let defaultProps: Props;

    beforeEach(() => {
        defaultProps = {
            name: "slider-test",
            accessible: "yes",
            style: [],
            valueAttribute: new EditableValueBuilder<Big>().withValue(new Big(140)).build(),
            editable: "default",
            minimumValue: dynamicValue<Big>(new Big(0)),
            maximumValue: dynamicValue<Big>(new Big(280)),
            stepSize: dynamicValue<Big>(new Big(1))
        };
    });

    it("renders", () => {
        const component = render(<Slider {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders no error while the value is resolving", () => {
        const component = render(
            <Slider {...defaultProps} valueAttribute={new EditableValueBuilder<Big>().isLoading().build()} />
        );
        expect(component.queryByTestId(`${defaultProps.name}-validation-message`)).toBeNull();
    });

    it("renders an error when the minimum is greater than the maximum", () => {
        const component = render(<Slider {...defaultProps} minimumValue={dynamicValue(new Big(300))} />);
        expect(component.queryByText("The minimum value can not be greater than the maximum value.")).not.toBeNull();
    });

    it("renders an error when the step size is negative", () => {
        const component = render(<Slider {...defaultProps} stepSize={dynamicValue(new Big(-10))} />);
        expect(component.queryByText("The step size can not be zero or less than zero.")).not.toBeNull();
    });

    it("renders an error when the step size is empty", () => {
        const component = render(<Slider {...defaultProps} stepSize={noValue} />);
        expect(component.queryByText("No step size provided.")).not.toBeNull();
    });

    it("renders an error when the minimum is empty", () => {
        const component = render(<Slider {...defaultProps} minimumValue={noValue} />);
        expect(component.queryByText("No minimum value provided.")).not.toBeNull();
    });

    it("renders an error when the maximum is empty", () => {
        const component = render(<Slider {...defaultProps} maximumValue={noValue} />);
        expect(component.queryByText("No maximum value provided.")).not.toBeNull();
    });

    it("renders an error when the minimum is equal to the maximum", () => {
        const value = new Big(10);
        const component = render(
            <Slider
                {...defaultProps}
                valueAttribute={new EditableValueBuilder<Big>().withValue(value).build()}
                minimumValue={dynamicValue(value)}
                maximumValue={dynamicValue(value)}
            />
        );
        expect(component.queryByText("The minimum value can not be equal to the maximum value.")).not.toBeNull();
    });

    it("renders an error when the value is less than the minimum", () => {
        const component = render(
            <Slider
                {...defaultProps}
                valueAttribute={new EditableValueBuilder<Big>().withValue(new Big(-50)).build()}
            />
        );
        expect(component.queryByText("The current value can not be less than the minimum value.")).not.toBeNull();
    });

    it("renders an error when the value is greater than the maximum", () => {
        const component = render(
            <Slider
                {...defaultProps}
                valueAttribute={new EditableValueBuilder<Big>().withValue(new Big(300)).build()}
            />
        );
        expect(component.queryByText("The current value can not be greater than the maximum value.")).not.toBeNull();
    });

    it("renders with the width of the parent view", () => {
        const component = render(
            <Slider
                {...defaultProps}
                style={[
                    {
                        container: { width: 100 },
                        highlight: {},
                        highlightDisabled: {},
                        marker: {},
                        markerActive: {},
                        markerDisabled: {},
                        track: {},
                        trackDisabled: {},
                        validationMessage: {}
                    }
                ]}
            />
        );
        fireEvent(component.getByTestId("slider-test"), "layout", { nativeEvent: { layout: { width: 100 } } });
        expect(component.getByTestId("slider-test").findByProps({ sliderLength: 100 })).not.toBeNull();
    });

    it("renders a validation message", () => {
        const value = new EditableValueBuilder<Big>().withValidation("Invalid").build();
        const component = render(<Slider {...defaultProps} valueAttribute={value} />);

        expect(component.queryByText("Invalid")).not.toBeNull();
    });

    it("handles an invalid step size", () => {
        const component = render(<Slider {...defaultProps} stepSize={dynamicValue(new Big(-10))} />);
        expect(component.getByTestId("slider-test").findByProps({ step: 1 })).not.toBeNull();
    });

    it("changes the value when swiping", () => {
        const onChangeAction = actionValue();
        const component = render(<Slider {...defaultProps} onChange={onChangeAction} />);

        fireEvent(getHandle(component), "responderGrant", { touchHistory: { touchBank: [] } });
        fireEvent(getHandle(component), "responderMove", responderMove(50));

        expect(onChangeAction.execute).not.toHaveBeenCalled();

        fireEvent(getHandle(component), "responderRelease", {});

        expect(defaultProps.valueAttribute.setValue).toHaveBeenCalledWith(new Big(190));
        expect(onChangeAction.execute).toHaveBeenCalledTimes(1);
    });

    it("does not change the value when non editable", () => {
        const onChangeAction = actionValue();
        const component = render(<Slider {...defaultProps} editable={"never"} onChange={onChangeAction} />);

        fireEvent(getHandle(component), "responderGrant", { touchHistory: { touchBank: [] } });
        fireEvent(getHandle(component), "responderMove", responderMove(50));
        fireEvent(getHandle(component), "responderRelease", {});

        expect(onChangeAction.execute).not.toHaveBeenCalled();
        expect(defaultProps.valueAttribute.setValue).not.toHaveBeenCalled();
    });
});

function getHandle(component: RenderAPI): ReactTestInstance {
    return component
        .getByTestId("slider-test")
        .findAllByType(View)
        .filter(instance => instance.props.onMoveShouldSetResponder)[0];
}

function responderMove(dx: number): any {
    return {
        touchHistory: {
            numberActiveTouches: 1,
            indexOfSingleActiveTouch: 0,
            touchBank: [
                {
                    touchActive: true,
                    currentTimeStamp: Date.now(),
                    currentPageX: dx,
                    currentPageY: 0,
                    previousPageX: 0,
                    previousPageY: 0
                }
            ]
        }
    };
}
