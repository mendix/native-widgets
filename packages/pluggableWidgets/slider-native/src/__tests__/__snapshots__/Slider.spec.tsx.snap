// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders 1`] = `
<View
  onLayout={[Function]}
  style={{}}
  testID="slider-test"
>
  <View
    style={
      [
        {
          "height": 50,
          "justifyContent": "center",
          "position": "relative",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "width": 280,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "#CECECE",
              "height": 2,
            },
            {
              "backgroundColor": "rgba(98,0,238, 0.3)",
            },
            {
              "backgroundColor": "rgb(98,0,238)",
            },
            {
              "width": 140,
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "backgroundColor": "#CECECE",
              "height": 2,
            },
            {
              "backgroundColor": "rgba(98,0,238, 0.3)",
            },
            undefined,
            {
              "width": 140,
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "height": 48,
              "justifyContent": "center",
              "position": "absolute",
              "width": 48,
            },
            {
              "left": 116,
              "top": -24,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "alignSelf": "stretch",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              },
              {
                "borderRadius": 15,
              },
            ]
          }
        >
          <View
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="slider-test$marker"
          >
            <View
              style={
                [
                  {
                    "backgroundColor": "#0D8675",
                    "borderRadius": 12,
                    "height": 12,
                    "width": 12,
                  },
                  [
                    undefined,
                    {
                      "backgroundColor": "rgb(98,0,238)",
                      "borderColor": "rgb(98,0,238)",
                    },
                  ],
                  undefined,
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
